### PR TITLE
Add OrderedSerialization2 support in Matrix2.

### PR DESCRIPTION
**Pull request by reconditesea**
To resolve issue #1418

### CODE REVIEW

**Comment by reconditesea** [comment]
@johnynek Is this good to go?

**Comment by johnynek** [comment]
Sorry. Didn't see this. Thanks for bringing to my attention. I'll review
Monday.

On Saturday, December 19, 2015, Kevin Lin <notifications@github.com> wrote:

> @johnynek <https://github.com/johnynek> Is this good to go?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/pull/1457#issuecomment-166011244>.
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin

**Comment by reconditesea** [comment]
Thanks, happy holidays!

File: scalding-core/src/main/scala/com/twitter/scalding/mathematics/Matrix2.scala [hunk 1]
```diff
@@ -18,6 +18,7 @@ package com.twitter.scalding.mathematics
 import cascading.flow.FlowDef
 import cascading.pipe.Pipe
 import cascading.tuple.Fields
+import com.twitter.scalding.serialization.{ OrderedSerialization, OrderedSerialization2 }
 import com.twitter.scalding.TDsl._
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/mathematics/Matrix2.scala]: can you remove TDsl while you are at it?

File: scalding-core/src/main/scala/com/twitter/scalding/mathematics/Matrix2.scala [hunk 2]
```diff
@@ -359,6 +359,13 @@ case class Product[R, C, C2, V](left: Matrix2[R, C, V],
 
   implicit override val rowOrd: Ordering[R] = left.rowOrd
   implicit override val colOrd: Ordering[C2] = right.colOrd
+  implicit def withOrderedSerialization: Ordering[(R, C2)] = {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/mathematics/Matrix2.scala]: Can we not repeat this code 3 times? What if we have this function in the companion object of `OrderedSerialization2` in this file:

https://github.com/twitter/scalding/blob/develop/scalding-serialization/src/main/scala/com/twitter/scalding/serialization/Serialization2.scala#L61

```scala
def maybeOrderedSerialization[A, B](implicit ordA: Ordering[A], ordB: Ordering[B]): Ordering[(A, B)] = ...
```

**Comment by johnynek** [comment]
Need a test for this that some Matrix2 job works with the option: https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/Config.scala#L118 `setRequiredOrderedSerialization(true)`.

In fact, we should write the test first and find an example that fails (I guess adding two matrices)

**Comment by reconditesea** [comment]
Moved the method implementation into OrderedSerialization2 companion object.
Will add new test cases shortly.

**Comment by reconditesea** [comment]
I had troubles in importing serialization test's intOrderedSerialization (https://github.com/twitter/scalding/blob/398545991c63019714d80ac22de6494adfd798f2/scalding-serialization/src/test/scala/com/twitter/scalding/serialization/SerializationProperties.scala#L46) into Matrix2 test, so pasted the code into Matrix2 test. 

BTW, feel like we can use slack channel for discussions like this more frequently.

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala [hunk 3]
```diff
@@ -40,6 +44,37 @@ class Matrix2Sum(args: Args) extends Job(args) {
...
+    def write(o: OutputStream, t: Int) = Try(Writer.write[Int](o, t))
+    def hash(t: Int) = t.hashCode
+    def compare(a: Int, b: Int) = java.lang.Integer.compare(a, b)
+    def compareBinary(a: InputStream, b: InputStream) =
+      OrderedSerialization.readThenCompare(a, b)(this)
+    val staticSize = Some(4)
+    def dynamicSize(i: Int) = staticSize
+  }
+
+  val p1: Pipe = Tsv("mat1", ('x1, 'y1, 'v1)).read
```
Comment (johnynek) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: this style is not recommended any more (.toTypedPipe), so let's just use TypedText.tsv the whole way:
```scala
val tp1 = TypedPipe.from(TypedText.tsv[(Int, Int, Double)]("mat1"))
```

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala [hunk 4]
```diff
@@ -293,6 +328,21 @@ class Matrix2Test extends WordSpec with Matchers {
     }
   }
 
+  "A MatrixSum job with Orderedserialization" should {
+    TUtil.printStack {
+      JobTest(new Matrix2SumOrderedSerialization(_))
```
Comment (johnynek) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: did you verify that this test fails if you don't have the intOS in scope?

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala [hunk 5]
```diff
@@ -40,6 +44,37 @@ class Matrix2Sum(args: Args) extends Job(args) {
   sum.write(TypedText.tsv[(Int, Int, Double)]("sum"))
 }
 
+class Matrix2SumOrderedSerialization(args: Args) extends Job(args) {
+
+  import Matrix2._
+  import cascading.pipe.Pipe
```
Comment (johnynek) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: let's not import Pipe, Fields, or TDsl.

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala [hunk 6]
```diff
@@ -40,6 +44,37 @@ class Matrix2Sum(args: Args) extends Job(args) {
...
+  import cascading.tuple.Fields
+  import com.twitter.scalding.TDsl._
+
+  override def config = super.config + (Config.ScaldingRequireOrderedSerialization -> "true")
+  implicit val intOS: OrderedSerialization[Int] = new OrderedSerialization[Int] {
```
Comment (johnynek) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: can you add a comment that this is just for a test, normally you would use: 
https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/serialization/RequiredBinaryComparators.scala#L21

```scala
import RequiredBinaryComparators.orderedSerialization
```

File: scalding-serialization/src/main/scala/com/twitter/scalding/serialization/Serialization2.scala [hunk 7]
```diff
@@ -58,6 +58,16 @@ class Serialization2[A, B](val serA: Serialization[A], val serB: Serialization[B
   } yield a + b
 }
 
+object OrderedSerialization2 {
+  def maybeOrderedSerialization[A, B](implicit ordA: Ordering[A], ordB: Ordering[B]): Ordering[(A, B)] = {
```
Comment (johnynek) [inline scalding-serialization/src/main/scala/com/twitter/scalding/serialization/Serialization2.scala]: can we actually rename this `maybeOrderedSerialization2`. Sorry, in scala you can't have implicits with colliding names (but you can have normal methods with colliding names as long as their parameters are different). We might want to add `maybe...[2-22]` at some point in the future.

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala (hunk 4 above)
Comment (reconditesea) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: Yup, it failed if I don't declaim intOS in scope.

Comment (johnynek) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: Cool. Just making sure. :)

Comment (reconditesea) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: Of source! :)
I will have another update to address other comments shortly.

File: scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala (hunk 6 above)
Comment (reconditesea) [inline scalding-core/src/test/scala/com/twitter/scalding/mathematics/Matrix2Test.scala]: I changed intOS as ```implicit val intOS = orderedSerialization[Int] ```.
I guess it's better to use the normal macro here.

**Pull request merged**